### PR TITLE
Due to changing behaviour of countElementsInTable(), drop sql in calls

### PR DESCRIPTION
**Pull request by Letargie**

### CODE REVIEW

File: inc/cartridgeitem_printermodel.class.php [hunk 1]
```diff
@@ -94,11 +94,9 @@ function getTabNameForItem(CommonGLPI $item, $withtemplate=0) {
    **/
    static function countForCartridgeItem(CartridgeItem $item) {
 
-      $restrict = "`".static::getTable()."`.`cartridgeitems_id` = '".$item->getField('id') ."'
-                   AND `".static::getTable()."`.`printermodels_id` = `glpi_printermodels`.`id`";
-
       return countElementsInTable(array('glpi_printermodels', static::getTable()),
-                                  $restrict);
+                                 [ static::getTable().cartridgeitems_id => $item->getField('id'),
```
Comment (remicollet) [inline inc/cartridgeitem_printermodel.class.php]: .'cartridgeitems_id'

File: inc/cartridgeitem_printermodel.class.php [hunk 2]
```diff
@@ -94,11 +94,9 @@ function getTabNameForItem(CommonGLPI $item, $withtemplate=0) {
...
+                    static::getTable().'printermodels_id' => 'glpi_printermodels'.'id']);
```
Comment (remicollet) [inline inc/cartridgeitem_printermodel.class.php]: 'glpi_printermodels.id'

File: inc/commondbtm.class.php [hunk 3]
```diff
@@ -1829,7 +1829,7 @@ function canUnrecurs() {
          $f = getForeignKeyFieldForTable($this->getTable());
 
          if (countElementsInTable($this->getTable(),
-                                  "`$f`='$ID' AND entities_id NOT IN $entities") > 0) {
+                                  [ $f => $ID, 'not' => [ 'entities_id' => $entities ]]) > 0) {
```
Comment (remicollet) [inline inc/commondbtm.class.php]: NOT in upercase.

Comment (remicollet) [inline inc/commondbtm.class.php]: entities is already a string (list), must be passed as an array

File: inc/commondbtm.class.php [hunk 4]
```diff
@@ -1847,16 +1847,14 @@ function canUnrecurs() {
                   if (is_array($field)) {
                      foreach ($field as $f) {
                         if (countElementsInTable($tablename,
-                                                 "`$f`='$ID'
-                                                   AND entities_id NOT IN $entities") > 0) {
+                                                 [ $f => $ID, 'not' => [ 'entities_id' => $entities ]]) > 0) {
```
Comment (remicollet) [inline inc/commondbtm.class.php]: idem

File: inc/commondbtm.class.php [hunk 5]
```diff
@@ -1847,16 +1847,14 @@ function canUnrecurs() {
...
                            return false;
                         }
                      }
 
                   } else {
                      if (countElementsInTable($tablename,
-                                              "`$field`='$ID'
-                                                AND entities_id NOT IN $entities") > 0) {
+                                              [ $field => $ID, 'not' => [ 'entities_id' => $entities ]]) > 0) {
```
Comment (remicollet) [inline inc/commondbtm.class.php]: idem

File: inc/commondbtm.class.php [hunk 6]
```diff
@@ -1883,13 +1881,10 @@ function canUnrecurs() {
...
-                                                           AND `$tablename`.`$typefield`
-                                                                  ='$itemtype'
-                                                           AND `$tablename`.`$devfield`
-                                                                  =`$itemtable`.id
-                                                           AND `$itemtable`.`entities_id`
-                                                                  NOT IN $entities") > '0') {
+                                                          [ $tablename.$field => $ID,
+                                                            $tablename.$typefield => $itemtype,
+                                                            $tablename.$devfield => $itemtable.'.id',
+                                                            'not' => [$itemtable.'.entities_id' => $entities ]]) > '0') {
```
Comment (remicollet) [inline inc/commondbtm.class.php]: Missing quote => "$tablename.$field"

File: inc/commondbtm.class.php [hunk 7]
```diff
@@ -1908,23 +1903,19 @@ function canUnrecurs() {
...
                               if (countElementsInTable(array($tablename, $othertable),
-                                                       "`$tablename`.`$field`=$ID
-                                                        AND `$tablename`.`$otherfield`
-                                                               =`$othertable`.id
-                                                        AND `$othertable`.`entities_id`
-                                                               NOT IN $entities") > '0') {
+                                                       [ $tablename.$field => $ID,
+                                                         $tablename.$otherfield => $othertable.'.id',
+                                                         'not' => [ $othertable.'.entities_id' => $entities ]]) > '0') {
                                  return false;
```
Comment (remicollet) [inline inc/commondbtm.class.php]: idem

File: inc/commondbtm.class.php [hunk 8]
```diff
@@ -1940,10 +1931,10 @@ function canUnrecurs() {
...
           && countElementsInTable(array('glpi_documents_items', 'glpi_documents'),
-                                  "`glpi_documents_items`.`items_id`='$ID'
-                                   AND `glpi_documents_items`.`itemtype`=".$this->getType()."
-                                   AND `glpi_documents_items`.`documents_id`=`glpi_documents`.`id`
-                                   AND `glpi_documents`.`entities_id` NOT IN $entities") > '0') {
+                                  ['glpi_documents_items.items_id'=> $ID,
+                                   'glpi_documents_items.itemtype'=> $this->getType(),
+                                   'glpi_documents_items.documents_id'=>'glpi_documents.id',
+                                   'not' => ['glpi_documents.entities_id' => $entities]]) > '0') {
          return false;
```
Comment (remicollet) [inline inc/commondbtm.class.php]: NOT

Comment (remicollet) [inline inc/commondbtm.class.php]: + $entities

File: inc/cartridgeitem_printermodel.class.php [hunk 9]
```diff
@@ -94,11 +94,9 @@ function getTabNameForItem(CommonGLPI $item, $withtemplate=0) {
    **/
    static function countForCartridgeItem(CartridgeItem $item) {
 
-      $restrict = "`".static::getTable()."`.`cartridgeitems_id` = '".$item->getField('id') ."'
-                   AND `".static::getTable()."`.`printermodels_id` = `glpi_printermodels`.`id`";
-
       return countElementsInTable(array('glpi_printermodels', static::getTable()),
-                                  $restrict);
+                                 [ static::getTable().'cartridgeitems_id' => $item->getField('id'),
+                    static::getTable().'printermodels_id' => 'glpi_printermodels.id']);
```
Comment (remicollet) [inline inc/cartridgeitem_printermodel.class.php]: Alignement + use 'FK' => [ static::getTable() => 'printermodels_id', 'glpi_printermodels" => 'id']

File: inc/commondbtm.class.php [hunk 10]
```diff
@@ -1817,19 +1817,15 @@ function canUnrecurs() {
          return true;
       }
 
-      $entities = "('".$this->fields['entities_id']."'";
-      foreach (getAncestorsOf("glpi_entities",$this->fields['entities_id']) as $papa) {
-         $entities .= ",'$papa'";
-      }
-
-      $entities .= ")";
+      $entities = getAncestorsOf("glpi_entities", $this->fields['entities_id']);
```
Comment (remicollet) [inline inc/commondbtm.class.php]: simple quote 'glpi_entities'

File: inc/commondbtm.class.php [hunk 11]
```diff
@@ -1883,13 +1877,10 @@ function canUnrecurs() {
...
-                                                          "`$tablename`.`$field`='$ID'
-                                                           AND `$tablename`.`$typefield`
-                                                                  ='$itemtype'
-                                                           AND `$tablename`.`$devfield`
-                                                                  =`$itemtable`.id
-                                                           AND `$itemtable`.`entities_id`
-                                                                  NOT IN $entities") > '0') {
+                                                          [ "$tablename.$field" => $ID,
+                                                            "$tablename.$typefield" => $itemtype,
+                                                            "$tablename.$devfield" => $itemtable.'.id',
```
Comment (remicollet) [inline inc/commondbtm.class.php]: FKEY

File: inc/commondbtm.class.php [hunk 12]
```diff
@@ -1908,23 +1899,19 @@ function canUnrecurs() {
                            if (is_array($rel[$tablename])) {
                               foreach ($rel[$tablename] as $otherfield) {
                                  if (countElementsInTable(array($tablename, $othertable),
-                                                          "`$tablename`.`$field`='$ID'
-                                                           AND `$tablename`.`$otherfield`
-                                                                  =`$othertable`.id
-                                                           AND `$othertable`.`entities_id`
-                                                                  NOT IN $entities") > '0') {
+                                                          [ "$tablename.$field" => $ID,
+                                                            "$tablename.$otherfield" => $othertable.'.id',
```
Comment (remicollet) [inline inc/commondbtm.class.php]: FKEY

File: inc/commondbtm.class.php [hunk 13]
```diff
@@ -1908,23 +1899,19 @@ function canUnrecurs() {
...
                            } else {
                               $otherfield = $rel[$tablename];
                               if (countElementsInTable(array($tablename, $othertable),
-                                                       "`$tablename`.`$field`=$ID
-                                                        AND `$tablename`.`$otherfield`
-                                                               =`$othertable`.id
-                                                        AND `$othertable`.`entities_id`
-                                                               NOT IN $entities") > '0') {
+                                                       [ "$tablename.$field" => $ID,
+                                                         "$tablename.$otherfield" => $othertable.'.id',
```
Comment (remicollet) [inline inc/commondbtm.class.php]: FKEY

File: inc/commondbtm.class.php [hunk 14]
```diff
@@ -1940,10 +1927,10 @@ function canUnrecurs() {
       // Doc links to this item
       if (($this->getType() > 0)
           && countElementsInTable(array('glpi_documents_items', 'glpi_documents'),
-                                  "`glpi_documents_items`.`items_id`='$ID'
-                                   AND `glpi_documents_items`.`itemtype`=".$this->getType()."
-                                   AND `glpi_documents_items`.`documents_id`=`glpi_documents`.`id`
-                                   AND `glpi_documents`.`entities_id` NOT IN $entities") > '0') {
+                                  ['glpi_documents_items.items_id'=> $ID,
+                                   'glpi_documents_items.itemtype'=> $this->getType(),
+                                   'glpi_documents_items.documents_id'=>'glpi_documents.id',
```
Comment (remicollet) [inline inc/commondbtm.class.php]: FKEY

File: inc/commondevice.class.php [hunk 15]
```diff
@@ -195,8 +195,8 @@ function canUnrecurs() {
             if ($item = getItemForItemtype($data["itemtype"])) {
                // For each itemtype which are entity dependant
                if ($item->isEntityAssign()) {
-                  if (countElementsInTable($itemtable, "id IN (".$data["ids"].")
-                                           AND entities_id NOT IN $entities") > 0) {
+                  if (countElementsInTable($itemtable, [ 'id' => $data["ids"],
+                                            'NOT' => ['entities_id' => $entities ]]) > 0) {
```
Comment (remicollet) [inline inc/commondevice.class.php]: check $entities (string vs array)

File: inc/item_project.class.php [hunk 16]
```diff
@@ -74,10 +74,9 @@ function getForbiddenStandardMassiveAction() {
    function prepareInputForAdd($input) {
 
       // Avoid duplicate entry
-      $restrict = " `projects_id` = '".$input['projects_id']."'
-                   AND `itemtype` = '".$input['itemtype']."'
-                   AND `items_id` = '".$input['items_id']."'";
-      if (countElementsInTable($this->getTable(), $restrict) > 0) {
+      if (countElementsInTable($this->getTable(), ['projects_id' => $input['projects_id'],
+                   'itemtype' => $input['itemtype'],
+                   'items_id' => $input['items_id']]) > 0) {
```
Comment (remicollet) [inline inc/item_project.class.php]: alignment

File: inc/networkequipment.class.php [hunk 17]
```diff
@@ -230,9 +230,9 @@ function canUnrecurs() {
                if ($item = getItemForItemtype($data["itemtype"])) {
                   // For each itemtype which are entity dependant
                   if ($item->isEntityAssign()) {
-                     if (countElementsInTable($itemtable, "id IN (".$data["ids"].")
-                                              AND entities_id NOT IN $entities") > 0) {
-                        return false;
+                     if (countElementsInTable($itemtable, ['id' => $data["ids"],
+                                               'NOT' => ['entities_id' => $entities ]]) > 0) {
```
Comment (remicollet) [inline inc/networkequipment.class.php]: check $entities (string vs array)

File: inc/printer.class.php [hunk 18]
```diff
@@ -155,8 +155,8 @@ function canUnrecurs() {
                   // For each itemtype which are entity dependant
                   if ($item->isEntityAssign()) {
 
-                     if (countElementsInTable($itemtable, "`id` IN (".$data["ids"].")
-                                              AND `entities_id` NOT IN $entities") > 0) {
+                     if (countElementsInTable($itemtable, ['id' => $data["ids"],
+                                              'NOT' => [ 'entities_id' => $entities]]) > 0) {
```
Comment (remicollet) [inline inc/printer.class.php]: $entities (string vs array)

File: inc/cartridgeitem_printermodel.class.php [hunk 19]
```diff
@@ -96,7 +96,7 @@ static function countForCartridgeItem(CartridgeItem $item) {
 
       return countElementsInTable(array('glpi_printermodels', static::getTable()),
                                         [static::getTable().'cartridgeitems_id' => $item->getField('id'),
-                    		         'FKEY' => [static::getTable() => 'printermodels_id', 'glpi_printermodels' => 'id']]);
+                                         'FKEY' => [static::getTable() => 'printermodels_id', 'glpi_printermodels' => 'id']]);
```
Comment (ddurieux) [inline inc/cartridgeitem_printermodel.class.php]: good ;)

File: inc/software.class.php [hunk 20]
```diff
@@ -217,7 +217,7 @@ static function updateValidityIndicator($ID) {
       if ($soft->getFromDB($ID)) {
          $valid = 1;
          if (countElementsInTable('glpi_softwarelicenses',
-                                  "`softwares_id`='$ID' AND NOT `is_valid`") > 0) {
+                                  ['softwares_id'=>$ID, 'not' => [ 'is_valid']]) > 0) {
```
Comment (remicollet) [inline inc/software.class.php]: NOT in upercase

**Comment by remicollet** [comment]
Can you please rebase to current master ?

**Comment by remicollet** [comment]
Merged, thanks.

**Pull request closed without merging**